Set YAMLFactory loader options to allow for larger files than the current SnakeYaml constraint (3MB)

fabiobrz opened the pull request:
Fixes #1351

This is to pass a `LoaderOptions`, that carries a hardcoded value for the limit of loaded files, to a [jackson-dataformats-text](https://github.com/FasterXML/jackson-dataformats-text) YAMLFactory instance [when OpenAPIProcessor creates it](https://github.com/smallrye/smallrye-open-api/blob/5d7ae7ce8c90ea454bf22f98ef13d30aa635d838/core/src/main/java/io/smallrye/openapi/runtime/io/OpenApiParser.java#L74).

Without this change the YAMLParser instance would use now the [SnakeYaml `ParserImpl` default (3 MB)](https://bitbucket.org/snakeyaml/snakeyaml/src/d9b0f480b1a63aca4678da7ab1915fcfc7d2a856/src/main/java/org/yaml/snakeyaml/LoaderOptions.java#lines-39) and fail when for instance loading static files that are larger than 3 MB:

```
at io.smallrye.openapi//io.smallrye.openapi.runtime.OpenApiProcessor.modelFromStaticFile(OpenApiProcessor.java:103)
	at org.wildfly.extension.microprofile.openapi-smallrye@28.0.0.Beta1-202301072046-b985cc92//org.wildfly.extension.microprofile.openapi.deployment.OpenAPIModelServiceConfigurator.get(OpenAPIModelServiceConfigurator.java:161)
	at org.wildfly.extension.microprofile.openapi-smallrye@28.0.0.Beta1-202301072046-b985cc92//org.wildfly.extension.microprofile.openapi.deployment.OpenAPIModelServiceConfigurator.get(OpenAPIModelServiceConfigurator.java:90)
	at org.wildfly.clustering.service@28.0.0.Beta1-202301072046-b985cc92//org.wildfly.clustering.service.FunctionalService.start(FunctionalService.java:63)
	... 8 more
Caused by: com.fasterxml.jackson.dataformat.yaml.JacksonYAMLParseException: The incoming YAML document exceeds the limit: 3145728 code points.
 at [Source: (FileInputStream); line: 109594, column: 25]
	at com.fasterxml.jackson.dataformat.jackson-dataformat-yaml@2.13.4//com.fasterxml.jackson.dataformat.yaml.YAMLParser.nextToken(YAMLParser.java:409)
	at com.fasterxml.jackson.core.jackson-core@2.13.4//com.fasterxml.jackson.core.JsonParser.nextFieldName(JsonParser.java:1038)
	at com.fasterxml.jackson.core.jackson-databind@2.13.4.2//com.fasterxml.jackson.databind.deser.std.BaseNodeDeserializer._deserializeContainerNoRecursion(JsonNodeDeserializer.java:440)
	at com.fasterxml.jackson.core.jackson-databind@2.13.4.2//com.fasterxml.jackson.databind.deser.std.JsonNodeDeserializer.deserialize(JsonNodeDeserializer.java:84)
	at com.fasterxml.jackson.core.jackson-databind@2.13.4.2//com.fasterxml.jackson.databind.deser.std.JsonNodeDeserializer.deserialize(JsonNodeDeserializer.java:20)
	at com.fasterxml.jackson.core.jackson-databind@2.13.4.2//com.fasterxml.jackson.databind.deser.DefaultDeserializationContext.readRootValue(DefaultDeserializationContext.java:323)
	at com.fasterxml.jackson.core.jackson-databind@2.13.4.2//com.fasterxml.jackson.databind.ObjectMapper._readTreeAndClose(ObjectMapper.java:4716)
	at com.fasterxml.jackson.core.jackson-databind@2.13.4.2//com.fasterxml.jackson.databind.ObjectMapper.readTree(ObjectMapper.java:3056)
	at io.smallrye.openapi//io.smallrye.openapi.runtime.io.OpenApiParser.parse(OpenApiParser.java:76)
	at io.smallrye.openapi//io.smallrye.openapi.runtime.OpenApiProcessor.modelFromStaticFile(OpenApiProcessor.java:101)
	... 11 more
Caused by: org.yaml.snakeyaml.error.YAMLException: The incoming YAML document exceeds the limit: 3145728 code points.
```

The PR can be improved by externalizing the limit value, instead than uising a 64 MB hardcoded value.

fabiobrz commented:
Hi @MikeEdgar - would you please have a look to this (maybe trivial) fix? 
I can improve it if you want, i.e. make the constraint value somehow configurable through OpenApiConfig, and/or add a test. 
Would be nice to have this in while WildFly 28 is ongoing.
WDYT?

fabiobrz commented:
And BTW I might have to target the 3.0.x branch as well, as it's the one WildFly is currently using: https://github.com/wildfly/wildfly/blob/main/pom.xml#L378

Anyway, let me know what your thoughts are and I'll move this one forward.

MikeEdgar commented:
@fabiobrz, thanks for the PR. I agree that a config value under `mp.openapi.extensions.smallrye.` would be useful for this.

fabiobrz commented:
> Thanks for your feedback @MikeEdgar! Will update this asap.

Hi @MikeEdgar - this is now updated and ready for review. I've been thinking a bit about adding a test, but don't have a strong opinion. WDYT?

fabiobrz commented:
> > Thanks for your feedback @MikeEdgar! Will update this asap.
> 
> Hi @MikeEdgar - this is now updated and ready for review. I've been thinking a bit about adding a test, but don't have a strong opinion. WDYT?

Thinking about this, I guess we do need tests, we've added a new vendor property and the behavior should be checked when this is set. Back to draft.

MikeEdgar commented:
> Thinking about this, I guess we do need tests, we've added a new vendor property and the behavior should be checked when this is set. Back to draft.

+1 , thanks @fabiobrz

MikeEdgar commented on core/src/test/java/io/smallrye/openapi/runtime/io/StaticFileConfigTest.java:
```diff
@@ -0,0 +1,23 @@
+    @Test
+    void testSettingMaximumFileSize() {
+        final Integer maximumFileSize = 8 * 1024 * 1024;
+        System.setProperty(OpenApiConstants.MAXIMUM_STATIC_FILE_SIZE, maximumFileSize.toString());
```
Maybe add a `System.clearProperty` in a `finally` block like the other test

fabiobrz replied:
You're right @MikeEdgar - Sure, I just missed it, thanks for catching it!

MikeEdgar reviewed (commented):
@fabiobrz , Looks pretty good to me, thanks. Just a small suggestion on one of the tests. wdyt?

MikeEdgar reviewed (approved):
LGTM, thanks @fabiobrz
